massa-deferred-calls/src/call.rs

```diff
@@ -0,0 +1,254 @@
+    pub(crate) address_serializer: AddressSerializer,
+    pub(crate) string_serializer: StringSerializer<U16VarIntSerializer, u16>,
+    pub(crate) vec_u8_serializer: VecU8Serializer,
+    pub(crate) amount_serializer: AmountSerializer,
+    pub(crate) u64_var_int_serializer: U64VarIntSerializer,
+    pub(crate) bool_serializer: BoolSerializer,
+}
+
+impl DeferredCallSerializer {
+    /// Serializes an `AsyncCall` into a `Vec<u8>`
+    pub fn new() -> Self {
+        Self {
+            slot_serializer: SlotSerializer::new(),
+            address_serializer: AddressSerializer::new(),
+            string_serializer: StringSerializer::new(U16VarIntSerializer::new()), // TODO: use max function name length
+            vec_u8_serializer: VecU8Serializer::new(), // TODO: use max parameters length
+            amount_serializer: AmountSerializer::new(),
+            u64_var_int_serializer: U64VarIntSerializer::new(),
+            bool_serializer: BoolSerializer::new(),
+        }
+    }
+}
+
+impl Serializer<DeferredCall> for DeferredCallSerializer {
+    fn serialize(&self, value: &DeferredCall, buffer: &mut Vec<u8>) -> Result<(), SerializeError> {
+        self.address_serializer
+            .serialize(&value.sender_address, buffer)?;
+        self.slot_serializer.serialize(&value.target_slot, buffer)?;
+        self.address_serializer
+            .serialize(&value.target_address, buffer)?;
+        self.string_serializer
+            .serialize(&value.target_function, buffer)?;
+        self.vec_u8_serializer
+            .serialize(&value.parameters, buffer)?;
+        self.amount_serializer.serialize(&value.coins, buffer)?;
+        self.u64_var_int_serializer
+            .serialize(&value.max_gas, buffer)?;
+        self.amount_serializer.serialize(&value.fee, buffer)?;
+        self.bool_serializer.serialize(&value.cancelled, buffer)?;
+        Ok(())
+    }
+}
+
+/// Deserializer for `AsyncCall`
+#[derive(Clone)]
+pub struct DeferredCallDeserializer {
+    slot_deserializer: SlotDeserializer,
+    address_deserializer: AddressDeserializer,
+    string_deserializer: StringDeserializer<U16VarIntDeserializer, u16>,
+    vec_u8_deserializer: VecU8Deserializer,
+    pub(crate) amount_deserializer: AmountDeserializer,
+    pub(crate) u64_var_int_deserializer: U64VarIntDeserializer,
+    bool_deserializer: BoolDeserializer,
+}
+
+impl DeferredCallDeserializer {
+    /// Deserializes a `Vec<u8>` into an `AsyncCall`
+    pub fn new(thread_count: u8) -> Self {
+        Self {
+            slot_deserializer: SlotDeserializer::new(
+                (Bound::Included(0), Bound::Included(u64::MAX)),
+                (Bound::Included(0), Bound::Excluded(thread_count)),
+            ),
+            address_deserializer: AddressDeserializer::new(),
+            string_deserializer: StringDeserializer::new(U16VarIntDeserializer::new(
+                Bound::Included(0),
+                Bound::Included(u16::MAX),
```

Comment:
Warning: do not use this to deserialize the function name because its length is limited by the config parameter `max_function_name_length`

Reply:
replaced